src/nbl/video/COpenGL_Swapchain.cpp

```diff
@@ -0,0 +1,414 @@
+
+#include "nbl/video/COpenGL_Swapchain.h"
+
+#include "nbl/video/surface/CSurfaceGL.h"
+#include "nbl/video/COpenGLSync.h"
+#include "nbl/video/IOpenGL_FunctionTable.h"
+#include "nbl/video/COpenGLFence.h"
+#include "nbl/video/COpenGLSemaphore.h"
+#include "nbl/video/COpenGLImage.h"
+#include "nbl/video/COpenGL_Queue.h"
+
+#include "nbl/system/IThreadHandler.h"
+
+namespace nbl::video
+{
+
+static inline constexpr uint32_t OpenGLFunctionTableSize = 6672u;
+using SThreadHandlerInternalState = std::array<uint8_t, OpenGLFunctionTableSize>;
+
+static_assert(OpenGLFunctionTableSize >= sizeof(COpenGLFunctionTable));
+static_assert(OpenGLFunctionTableSize >= sizeof(COpenGLESFunctionTable));
+
+IOpenGL_FunctionTable* getFunctionPointer(video::E_API_TYPE apiType, SThreadHandlerInternalState* internalState)
+{
+    if (apiType == video::EAT_OPENGL)
+        return reinterpret_cast<COpenGLFunctionTable*>(internalState);
+    else if (apiType == video::EAT_OPENGL_ES)
+        return reinterpret_cast<COpenGLESFunctionTable*>(internalState);
+    else assert(false);
+}
+
+class COpenGL_SwapchainThreadHandler final : public system::IThreadHandler<COpenGL_SwapchainThreadHandler, SThreadHandlerInternalState>
+{
+    friend COpenGLSwapchain;
+    friend COpenGLESSwapchain;
+    static inline constexpr uint32_t MaxImages = 4u;
+    using base_t = system::IThreadHandler<COpenGL_SwapchainThreadHandler, SThreadHandlerInternalState>;
+    friend base_t;
+
+public:
+    COpenGL_SwapchainThreadHandler(const egl::CEGL* _egl,
+        core::smart_refctd_ptr<IOpenGL_LogicalDevice> dev,
```

Comment:
you can use dumb pointer, the handler cannot quit/destroy before the swapchain itself